pygsp/graphs/fourier.py

```diff
@@ -88,26 +103,49 @@ def compute_fourier_basis(self, recompute=False):
         True
 
         """
+        if n_eigenvectors is None:
+            n_eigenvectors = self.N
 
-        if hasattr(self, '_e') and hasattr(self, '_U') and not recompute:
+        if hasattr(self, '_e') and hasattr(self, '_U') and not recompute \
+                and n_eigenvectors <= len(self.e):
             return
 
         assert self.L.shape == (self.N, self.N)
-        if self.N > 3000:
+        if self.N > 3000 and n_eigenvectors == self.N:
             self.logger.warning('Computing the full eigendecomposition of a '
                                 'large matrix ({0} x {0}) may take some '
-                                'time.'.format(self.N))
-
-        # TODO: handle non-symmetric Laplatians. Test lap_type?
-
-        self._e, self._U = np.linalg.eigh(self.L.toarray())
-        # Columns are eigenvectors. Sorted in ascending eigenvalue order.
-
-        # Smallest eigenvalue should be zero: correct numerical errors.
-        # Eigensolver might sometimes return small negative values, which
-        # filter's implementations may not anticipate. Better for plotting too.
-        assert -1e-12 < self._e[0] < 1e-12
-        self._e[0] = 0
+                                'time. Consider computing the partial '
+                                'eigendecomposition with '
+                                'n_eigenvectors=n'.format(self.N))
+        elif self.N * n_eigenvectors > 3000**2:
+            self.logger.warning('Computing the eigendecomposition of a '
+                                'large matrix ({0} x {0}) with {1} '
+                                'eigenvectors may take some time. Consider '
+                                'decreasing n_eigenvectors'.format(
+                                    self.N, n_eigenvectors))
+
+        # TODO: handle non-symmetric Laplacians. Test lap_type?
+        if n_eigenvectors == self.N:
+            self._e, self._U = np.linalg.eigh(self.L.toarray())
+            # Columns are eigenvectors. Sorted in ascending eigenvalue order.
+
+            # Smallest eigenvalue should be zero: correct numerical errors.
+            # Eigensolver might sometimes return small negative values, which
+            # filter's implementations may not anticipate. Better for plotting
+            # too.
+            assert -1e-12 < self._e[0] < 1e-12
+            self._e[0] = 0
+        else:
+            # fast partial eigendecomposition of hermitian matrices
+            # this will be slow if L is non-symmetric
+            if np.sum(self.L != self.L.T) != 0:
```

Comment:
I don't think we should test for symmetry. It's assumed (in the current implementation at least) that `L` is symmetric. We only support the combinatorial and normalized Laplacian for now. If at some point we had support for non-symmetric laplacians, we'll have to change many things (and we'll test the `lap_type`). As we're using the hermitian version of the eigensolvers, only the lower or upper triangular part of the matrix will be considered anyway.